simple_network_sim/network_of_populations.py

```diff
@@ -231,38 +236,72 @@ def getTotalSuscept(nodeState):
-    :type newInfections: int
+    :param newContacts: The number of new contacts to be distributed across age ranges.
+    :type newContacts: int
+    :param stochastic: Whether to run the model in a stochastic or deterministic mode
+    :type stochastic: bool
     :return: The number of new infections in each age group.
     :rtype: A dictionary of ages (keys) and the number of new infections (values)
     """
     ageToSus = {}
-    newInfectionsByAge = {}
     totalSus = 0.0
     for age in getAges(nodeState):
         sus = getSusceptibles(age, nodeState)
         ageToSus[age] = sus
         totalSus += sus
-    if totalSus < newInfections:
-        logger.error("totalSus < incoming infectious (%s < %s) - adjusting to totalSus", totalSus, newInfections)
-        newInfections = totalSus
-    for age, sus in ageToSus.items():
-        if totalSus > 0:
-            newInfectionsByAge[age] = (sus / totalSus) * newInfections
-        else:
-            newInfectionsByAge[age] = 0.0
+
+    if totalSus < newContacts:
+        logger.error("totalSus < incoming contacts (%s < %s) - adjusting to totalSus", totalSus, newContacts)
+        newContacts = totalSus
+
+    if totalSus > 0:
+        newInfectionsByAge = _distributeContactsOverAges(ageToSus, totalSus, newContacts, stochastic)
+    else:
+        newInfectionsByAge = {age: 0. for age in ageToSus}
+
     return newInfectionsByAge
 
 
-def getIncomingInfectiousContactsByNode(graph, currentState, movementMultiplier, infectiousStates):
+def _distributeContactsOverAges(ageToSusceptibles, totalSusceptibles, newContacts, stochastic):
+    """Distribute the number of new contacts across a region. This function distributes
+    newContacts among the susceptibles according to the relative size of each age group.
+    Two versions are possible:
+    1) Deterministic
+    We simply allocate newContacts by proportion of people in each age groups
+    2) Stochastic
+    newContacts are sampled randomly from the population from a Multinomial distribution,
+    with: k=newContacts, p=proportion of people in age groups
+
+    :param ageToSusceptibles: The number of susceptibles in each age group
+    :type ageToSusceptibles: A dictionary with age as keys and the number of susceptible
+    individuals as values.
+    :param totalSusceptibles: Total number of susceptible people in all age groups
+    :type totalSusceptibles: float
+    :param newContacts: The number of new contacts to be distributed across age ranges.
+    :type newContacts: int
+    :param stochastic: Whether to run the model in a stochastic or deterministic mode
+    :type stochastic: bool
+    :return: The number of new infections in each age group.
+    :rtype: A dictionary of ages (keys) and the number of new infections (values)
+    """
+    if stochastic:
+        ageProbabilities = np.array(list(ageToSusceptibles.values())) / totalSusceptibles
+        allocationsByAge = stats.multinomial.rvs(newContacts, ageProbabilities)
```

Comment:
I don't have much experience with the scipy.stats module. But all functions seem to take a random number generator as a paramter. Could we store the generator in the NetworkOfPopulations tuple and use the same generator everywhere? I think the seed used needs to be a paramter or an output of the model as well (so that people can reproduce runs). I think that will affect the way we parelelise code too, but, even if we run it across different machines, we can still prepare a set of seeds ahead of time.

Reply:
I agree with you that we need to store the seed for reproducibility. However, there are many pitfalls here. 
1) We need a different seed for every trial (if we use the same seed for every trial every trial will be identical). Using random seeds is not good either, as they might still be repeated out of pure luck. Usually using sequential seeds is the way to go (see https://www.johndcook.com/blog/2016/01/29/random-number-generator-seed-mistakes/)
2) Randomness is not guaranteed for numbers generated by different seeds! Only for number generated sequentially out of a single seed (see e.g. https://blogs.unity3d.com/2015/01/07/a-primer-on-repeatable-random-numbers/). So ideally we set the seed once and for all before doing all the trials SEQUENTIALLY.

For parallelizing the random trials, I will need to think a bit about it.

Reply:
Regarding passing the same generator to all methods, that's why I suggested putitng it in the NetworkOfPopulations tuple. My expectation is that we are going to parallelise the multiple runs of the model, not internally (I've did a few attempls of internal parallelisation and that didn't really to help, it's tricky to avoid copying a bunch of data around). So, we should always consider the whole model is always going to be single threaded.

Now, generating multiple seeds for each run, I don't know what's the best way to do it. Don't we just need to guarantee that each run has a different seed value?

Reply:
Thanks, sounds good for the single threaded model. I've just made a commit to initialize a random_state generator at network creation.

For multiple seeds on each run, theoretically it's not quite correct (as per my second link value, they show an example of a series of n numbers from one seed vs 1 number from n seeds). But in practice lot of places do like this. I'll try to look around to see how it's usually done.

Reply:
Maybe I'm misunderstanding something, but the blog post you pasted in the first link said exaclty what I said:

> In this case it would have been better to seed each process with sequential seeds: give the first process seed 1, the second seed 2, etc. The seeds don’t have to be random; they just have to be unique. If you’re using a good random number generator, the outputs of 1,000 processes seeded with 1, 2, 3, …, 1000 will be independent.

That's our scenario. The only reason I wanted to use uuids or something like that is because we want the ability of running different simulations, rather than always use the same fixed seeds. Although always keeping the same seeds may be ok too (as we don't expect much difference between runs of thousands of concurrent models).

Reply:
Yes sorry for not more clear, that's what most practitioners do (and most places I know of/been to). But in theory this isn't correct (the second blog post shows that), as random numbers generated from different seeds are not independent from each other. So I think we can start by giving sequential seeds is fine for now, but I will check how that problem is usually solved otherwise.

Reply:
I think I understand your point now. I was considering different runs as independent runs. In that case, we wouldn't care that two random sequences are not independent from each other. But they are not actually independent because we are going to average out in the end, right? I think to get around that problem we'd have to either write or find a very sophisticated random number generator which is able to create multiple independent sequences.

Reply:
Yes exactly, we average things that are not quite independent so the expectation will be slightly off. But given it seems it's a problem that half the world would have, even if we don't do anything special about it in the end, we should keep in mind and have a bit of a look around at least. 

I've known about that fact for some time but never knew if there was a solution to that.